pkg/nfd-client/topology-updater/nfd-topology-updater.go

```diff
@@ -0,0 +1,240 @@
+	"golang.org/x/net/context"
+	"sigs.k8s.io/node-feature-discovery/pkg/apihelper"
+	nfdclient "sigs.k8s.io/node-feature-discovery/pkg/nfd-client"
+	"sigs.k8s.io/node-feature-discovery/pkg/podres"
+	"sigs.k8s.io/node-feature-discovery/pkg/resourcemonitor"
+	pb "sigs.k8s.io/node-feature-discovery/pkg/topologyupdater"
+	"sigs.k8s.io/node-feature-discovery/pkg/utils"
+	"sigs.k8s.io/node-feature-discovery/pkg/version"
+)
+
+// Command line arguments
+type Args struct {
+	nfdclient.Args
+	NoPublish      bool
+	Oneshot        bool
+	KubeConfigFile string
+}
+
+type NfdTopologyUpdater interface {
+	nfdclient.NfdClient
+	Update(v1alpha1.ZoneList) error
+}
+
+type staticNodeInfo struct {
+	args                Args
+	resourcemonitorArgs resourcemonitor.Args
+	tmPolicy            string
+}
+
+type nfdTopologyUpdater struct {
+	nfdclient.NfdBaseClient
+	nodeInfo  *staticNodeInfo
+	certWatch *utils.FsWatcher
+	client    pb.NodeTopologyClient
+	stop      chan struct{} // channel for signaling stop
+}
+
+// Create new NewTopologyUpdater instance.
+func NewTopologyUpdater(args *Args, resourcemonitorArgs *resourcemonitor.Args, policy string) (NfdTopologyUpdater, error) {
+	base, err := nfdclient.NewNfdBaseClient(&args.Args)
+	if err != nil {
+		return nil, err
+	}
+
+	nfd := &nfdTopologyUpdater{
+		NfdBaseClient: base,
+		nodeInfo: &staticNodeInfo{
+			args:                *args,
+			resourcemonitorArgs: *resourcemonitorArgs,
+			tmPolicy:            policy,
+		},
+		stop: make(chan struct{}, 1),
+	}
+	return nfd, nil
+}
+
+// Run nfdTopologyUpdater client. Returns if a fatal error is encountered, or, after
+// one request if OneShot is set to 'true' in the updater args.
+func (w *nfdTopologyUpdater) Run() error {
+	klog.Infof("Node Feature Discovery Topology Updater %s", version.Get())
+	klog.Infof("NodeName: '%s'", nfdclient.NodeName)
+
+	podResClient, err := podres.GetPodResClient(w.nodeInfo.resourcemonitorArgs.PodResourceSocketPath)
+	if err != nil {
+		klog.Fatalf("Failed to get PodResource Client: %v", err)
+	}
+
+	kubeApihelper := apihelper.K8sHelpers{Kubeconfig: w.nodeInfo.args.KubeConfigFile}
+	var resScan resourcemonitor.ResourcesScanner
+
+	resScan, err = resourcemonitor.NewPodResourcesScanner(w.nodeInfo.resourcemonitorArgs.Namespace, podResClient, kubeApihelper)
+	if err != nil {
+		klog.Fatalf("Failed to initialize ResourceMonitor instance: %v", err)
+	}
+
+	// CAUTION: these resources are expected to change rarely - if ever.
+	// So we are intentionally do this once during the process lifecycle.
+	// TODO: Obtain node resources dynamically from the podresource API
+	// zonesChannel := make(chan v1alpha1.ZoneList)
+	var zones v1alpha1.ZoneList
+
+	resAggr, err := resourcemonitor.NewResourcesAggregator(w.nodeInfo.resourcemonitorArgs.FsRoot, podResClient)
+	if err != nil {
+		klog.Fatalf("Failed to obtain node resource information: %v", err)
+	}
+
+	klog.Infof("resAggr is: %v\n", resAggr)
```

Comment:
If err != nil, this line will crash bc theres no `resAggr`

Reply:
Ack

Reply:
Sure, this has been updated.